### PR TITLE
web-wallet: Integration with latest wallet-js and Sync improvements

### LINKED ISSUE #1561: web-wallet: Add Current Block Height step in Create Wallet flow
#### Summary
Implement an additional step in the Create Wallet flow that displays the current block height.

[Figma](https://www.figma.com/file/ImsgL6Xf0R31HFUIw2xxnw/Wallet-Extension-(FEB-08)?type=design&node-id=3544%3A80137&mode=design&t=6v58SKaSJrVz2B0M-1)

1st setup

#### Related task:
- [x] https://github.com/dusk-network/dusk-wallet-js/issues/75

### LINKED ISSUE #1567: web-wallet: Ensure newly created Wallet does not sync from genesis
#### Summary

Currently, creating a Wallet takes a long time, dependent on the blockchain size. We should optimize performance by allowing users to sync from the point when the user's wallet has been created (which means no initial sync is needed).

#### Related Tasks

- [x] https://github.com/dusk-network/dusk-wallet-js/issues/77

### LINKED ISSUE #1568: web-wallet: Add Syncing Settings in the Reset Wallet flow
#### Summary
Implement an additional step in the Reset Wallet flow that lets the user choose between syncing from genesis or from a specific block height.

#### Possible solution design or implementation
[Figma](https://www.figma.com/file/ImsgL6Xf0R31HFUIw2xxnw/Wallet-Extension-(FEB-08)?type=design&node-id=2784%3A21122&mode=design&t=Y4kpsbyvkPoYo2cz-1)

sync

sync-1

**Note:** This should be part of the Wizard, as an additional step. The Start Sync button is not required either.

#### Related Tasks

- [x] https://github.com/dusk-network/dusk-wallet-js/issues/77

### LINKED ISSUE #1570: web-wallet: Update the Sync indicator to reflect the actual progress
#### Summary

Currently the sync is an asynchronous operation that can take a lot of time, especially if synced from genesis (block 0); but there is no indicator whatsoever about its progression.

We should provide an easy way to communicate the progression of the synchronization.

#### Related Tasks

- [x] https://github.com/dusk-network/dusk-wallet-js/issues/78

### LINKED ISSUE #1595: Update Wallet service to use `dusk-wallet-js` 0.5.2
#### Summary

- Expose function to retrieve the current network block height
- Add possibility to sync from a specific block height

### PR DESCRIPTION
**Pull request by nortonandreev**
Resolves #1561, Resolves #1567, Resolves #1568, Resolves #1570, Resolves #1595

This PR includes the following features / enhancements:

- Integration with the latest wallet-js library
- Display of the current network block height during wallet creation
- Prevention of full sync for newly created wallets
- Option for users to restore a wallet from either genesis or a specific block height
- Relocation of the initial sync to the wallet restore flow
- A sync indicator that displays progress for both initial and subsequent syncs

### CODE REVIEW

File: web-wallet/vite.config.js [hunk 1]
```diff
@@ -65,6 +65,9 @@ export default defineConfig(({ mode }) => {
     plugins:
       mode === "development" ? [basicSsl(), ...commonPlugins] : commonPlugins,
     server: {
+      fs: {
+        allow: ["./lib"],
+      },
```
Comment (nortonandreev) [inline web-wallet/vite.config.js:70]: temp while serving the dusk-wallet-js locally

File: web-wallet/package.json [hunk 2]
```diff
@@ -35,7 +35,7 @@
   "type": "module",
   "version": "0.5.0",
   "dependencies": {
-    "@dusk-network/dusk-wallet-js": "0.4.2",
+    "@dusk-network/dusk-wallet-js": "file:./lib/dusk-wallet-js",
```
Comment (nortonandreev) [inline web-wallet/package.json:38]: Temp until the wallet js update is released – should not be merged before then.

**Comment by nortonandreev** [comment]
The `should initialize the wallet without setting a password` (in Wallet Create) fails for some reason – seems that the userId is still set in the settings store. If I call `settingsStore.reset()` explicitly before the Create component is rendered, it works as expected. However, it worries me that something is wrong with the store resetting/setting.

Could you please assist debugging, @ascartabelli? 🙏 Thanks!

File: web-wallet/.eslintignore [hunk 3]
```diff
@@ -4,10 +4,15 @@ node_modules
 /.svelte-kit
 /package
 .env
-.env.*
+.env.\*
```
Comment (ascartabelli) [inline web-wallet/.eslintignore:7]: Why the added escaping here?

File: web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  export let displayValue;
+
+  /** @type {string} */
+  export let rawValue;
+
+  /** @type {boolean} */
+  export let disabled = false;
+
+  function copyToClipboard() {
+    navigator.clipboard.writeText(rawValue);
```
Comment (ascartabelli) [inline web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte:19]: As the `writeText` returns a promise, we should show the success toast if it settles without errors (and and error toast if it fails). 

```js
navigator.clipboard.writeText(newValue).then(...).catch(...)
```

File: web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte
Comment (ascartabelli) [inline web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte:1]: This component can probably be used elsewhere (maybe in a separate issue).

File: web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  />
+</div>
+
+<style lang="postcss">
+  .copy-field {
+    display: flex;
+    align-items: center;
+    gap: var(--default-gap);
+  }
+  :global {
```
Comment (ascartabelli) [inline web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte:47]: why not make the whole set of rules global?

File: web-wallet/src/lib/dusk/number/createNumberFormatter.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Creates a locale aware number formatter
+ *
+ * @param {String} locale A BCP 47 language tag
+ * @param {Number|Undefined} digits The minimum fraction digits that should display
```
Comment (ascartabelli) [inline web-wallet/src/lib/dusk/number/createNumberFormatter.js:5]: You say "minimum" her, but the the digits are set to `maximumFractionDigits` in the options.

File: web-wallet/src/lib/stores/walletStore.js [hunk 7]
```diff
@@ -55,17 +48,35 @@ const getCurrentAddress = () => get(walletStore).currentAddress;
...
+async function updateAfterSync() {
+  const store = get(walletStore);
+
+  // @ts-expect-error
+  const balance = await walletInstance.getBalance(store.currentAddress);
+
+  set({
+    ...store,
+    balance,
+    syncStatus: { isInProgress: false, current: 0, last: 0, error: null },
```
Comment (ascartabelli) [inline web-wallet/src/lib/stores/walletStore.js:60]: Maybe it's clearer if we set it to `initialState.syncStatus` instead?

File: web-wallet/src/lib/stores/walletStore.js [hunk 8]
```diff
@@ -55,17 +48,35 @@ const getCurrentAddress = () => get(walletStore).currentAddress;
...
-/** @type {() => Promise<void>} */
+/** @type {WalletStoreServices["clearLocalData"]} */
 const clearLocalData = async () => walletInstance?.reset();
 
-/** @type {(wallet: Wallet) => Promise<void>} */
-const clearLocalDataAndInit = (wallet) =>
-  wallet.reset().then(() => init(wallet));
+/** @type {WalletStoreServices["clearLocalDataAndInit"]} */
+const clearLocalDataAndInit = async (wallet, syncFromBlock) => {
+  await wallet.reset();
```
Comment (ascartabelli) [inline web-wallet/src/lib/stores/walletStore.js:74]: Why this change?
Isn't better to just return the "reset & init promise" as before?

File: web-wallet/src/lib/stores/walletStore.js [hunk 9]
```diff
@@ -114,54 +112,89 @@ async function init(wallet) {
...
+        error: null,
+      },
+    });
 
     syncController = new AbortController();
+
+    const syncOptions = {
+      from,
+      signal: syncController.signal,
+      onblock: (/** @type {number} */ current, /** @type {number} */ last) => {
```
Comment (ascartabelli) [inline web-wallet/src/lib/stores/walletStore.js:167]: Maybe it would be clearer by adding a `@type {(current: number, last: number) => void}` before instead of mixing code and comments?

File: web-wallet/src/routes/(welcome)/setup/restore/NetworkSyncProgress.svelte [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  import { walletStore } from "$lib/stores";
+  import { mdiCubeOutline } from "@mdi/js";
+  import { onDestroy } from "svelte";
+
+  $: ({ syncStatus } = $walletStore);
+
+  /** @type {boolean} */
+  export let isValid = false;
+
+  let syncStarted = false;
```
Comment (ascartabelli) [inline web-wallet/src/routes/(welcome)/setup/restore/NetworkSyncProgress.svelte:13]: Do we really need this extra flag?

**Review by ascartabelli — COMMENTED** [review]
This was just a quick pre-review, left some comments around.

In any case: really a great job!

File: web-wallet/.eslintignore (hunk 3 above)
Comment (nortonandreev) [reply web-wallet/.eslintignore:7]: Not too sure. Is the `.env.` needed at all, or can we just leave the `.env` on line 6?

File: web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte (hunk 4 above)
Comment (nortonandreev) [reply web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte:19]: Fixed.

File: web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte
Comment (nortonandreev) [reply web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte:1]: Yeah, this is why it's under the `dusk/components` path. I will also add a test.

File: web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte (hunk 5 above)
Comment (nortonandreev) [reply web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte:47]: I have moved it to a CopyField.css file, similarly to how we have it for other components.

File: web-wallet/src/routes/(welcome)/setup/restore/NetworkSyncProgress.svelte (hunk 10 above)
Comment (nortonandreev) [reply web-wallet/src/routes/(welcome)/setup/restore/NetworkSyncProgress.svelte:13]: Screenshot 2024-08-26 at 14 13 21
Screenshot 2024-08-26 at 14 13 26


Without it, it appears as if the sync has completed for a split second, before the sync actually starts.

File: web-wallet/src/lib/stores/walletStore.js (hunk 8 above)
Comment (nortonandreev) [reply web-wallet/src/lib/stores/walletStore.js:74]: Will revert the change. I guess I have been experimenting with the function.

File: web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte
Comment (nortonandreev) [reply web-wallet/src/lib/dusk/components/CopyField/CopyField.svelte:1]: Edit: Have moved as an app component, as suggested.

**Review by ascartabelli — APPROVED** [review]
